### PR TITLE
Find duplicates from constraints.txt

### PR DESCRIPTION
**Pull request by rd4398**
In this commit, I have created a helper to process duplicate entries in constraints.txt file and shift those to the start of the file as per discussion in https://github.com/python-wheel-build/fromager/issues/151

### CODE REVIEW

**Review by shubhbapna — COMMENTED** [review]
You might get away with just sorting the `_build_stack` list actually so something like this:
```python
sorted(self._build_stack, lambda item: item["dist"])
```
This should end up grouping all the items which have the same dist together

File: src/fromager/context.py [hunk 1]
```diff
@@ -119,13 +119,40 @@ def add_to_build_order(
             # converted to JSON without help.
             json.dump(self._build_stack, f, indent=2, default=str)
         with open(self._constraints_filename, "w") as f:
-            for step in self._build_stack:
+            for step in self._process_constraints(self._build_stack):
                 comment = " ".join(
                     f"-{dep_type}-> {req.name}({version})"
                     for dep_type, req, version in step["why"]
                 )
                 f.write(f'{step["dist"]}=={step["version"]}  # {comment}\n')
```
Comment (shubhbapna) [inline src/fromager/context.py:127]: After processing the build_stack wouldn't we have to comment out any duplicates?

Comment (rd4398) [reply src/fromager/context.py:127]: I discussed this  with Doug, and he said `Maybe we don't pick for now, and just list the ones with duplicates at the top of the file so it's easy for a person to pick them`

Comment (rd4398) [reply src/fromager/context.py:127]: I tried sorted() but we want all the duplicates at the top of file followed by non duplicates and sorted doesn't do that even though it groups items

Comment (shubhbapna) [reply src/fromager/context.py:127]: Oh so you want any pkg with duplicates to be on the top of the file not just grouped together in the file?

Comment (rd4398) [reply src/fromager/context.py:127]: Yeah!

Comment (shubhbapna) [reply src/fromager/context.py:127]: You could try this:
```python
# count the number of occurrences of a given dist
count = collections.Counter([item["dist"] for item in self._build_stack])
# first group all same dist items together as it will help with handling cases when 2 dists have the same count
grouped_build_stack = sorted(input, key=lambda item: item["dist"])
# sort based on the count
sorted(grouped_build_stack, key=lambda item: count[item["dist"]], reverse=True)
```

Comment (rd4398) [reply src/fromager/context.py:127]: Okay, let me try this

File: tests/test_context.py [hunk 2]
```diff
@@ -166,3 +166,80 @@ def test_build_order_name_canonicalization(tmp_context):
         },
     ]
     assert expected == contents
+
+
+def test_processing_constraints_file(tmp_context):
```
Comment (shubhbapna) [inline tests/test_context.py:171]: I would add more edge cases since we already found one that didn't work before

Comment (rd4398) [reply tests/test_context.py:171]: Okay, I will do that!
Also, I am not able to understand why e2e tests are failing

**Review by shubhbapna — CHANGES_REQUESTED** [review]
You can separate out the one big unit test into individual test cases that describes what kind of edge cases you are testing

**Comment by rd4398** [comment]
Okay, I will do that

**Comment by dhellmann** [comment]
It looks like @shubhbapna has the code review well in hand. My only feedback is to please squash the commits in your branch before the final review and merge.

**Review by shubhbapna — APPROVED** [review]
LGTM!
